webapp/src/components/common/GroupedDataTable/index.tsx

```diff
@@ -1,69 +1,299 @@
+              ? !rowIds.some((id) => prev[id]) // Select/Deselect all
+              : !prev[row.id];
+
+            return {
+              ...prev,
+              ...rowIds.reduce((acc, id) => ({ ...acc, [id]: newValue }), {}),
+            };
+          });
+        },
+        selected: rowSelection[row.id],
+        sx: { cursor: isPending ? "wait" : "pointer" },
+      };
+    },
+    // Toolbars
+    renderTopToolbarCustomActions: ({ table }) => (
+      <Box sx={{ display: "flex", gap: 1 }}>
+        {onCreate && (
+          <Button
+            startIcon={<AddCircleOutlineIcon />}
+            variant="contained"
+            size="small"
+            onClick={() => setOpenDialog("add")}
+          >
+            {t("button.add")}
+          </Button>
+        )}
+        {onDuplicate && (
+          <Button
+            startIcon={<ContentCopyIcon />}
+            variant="outlined"
+            size="small"
+            onClick={() => setOpenDialog("duplicate")}
+            disabled={table.getSelectedRowModel().rows.length !== 1}
+          >
+            {t("global.duplicate")}
+          </Button>
+        )}
+        {onDelete && (
+          <Button
+            startIcon={<DeleteOutlineIcon />}
+            color="error"
+            variant="outlined"
+            size="small"
+            onClick={() => setOpenDialog("delete")}
+            disabled={table.getSelectedRowModel().rows.length === 0}
+          >
+            {t("global.delete")}
+          </Button>
+        )}
+      </Box>
+    ),
+    renderToolbarInternalActions: ({ table }) => (
+      <>
+        <MRT_ToggleGlobalFilterButton table={table} />
+        <MRT_ToggleFiltersButton table={table} />
+      </>
+    ),
+    onRowSelectionChange: setRowSelection,
+    // Styles
+    muiTablePaperProps: { sx: { display: "flex", flexDirection: "column" } }, // Allow to have scroll
+    ...R.mergeDeepRight(getTableOptionsForAlign("right"), {
+      muiTableBodyCellProps: {
+        sx: { borderBottom: "1px solid rgba(224, 224, 224, 0.3)" },
+      },
+    }),
+  });
+
+  const selectedRows = table
+    .getSelectedRowModel()
+    .rows.map((row) => row.original);
+  const selectedRow = selectedRows.length === 1 ? selectedRows[0] : null;
+
+  ////////////////////////////////////////////////////////////////
+  // Optimistic
+  ////////////////////////////////////////////////////////////////
+
+  const addPendingRow = (row: TRow<TGroups[number]>) => {
+    const pendingRow = fillPendingRow?.(row) || row;
+
+    pendingRows.current.push(pendingRow);
+
+    // Type can be asserted as `TData` because the row will be checked in cell renders
+    // and `fillPendingRow` allows to add needed data
+    setTableData((prev) => [...prev, pendingRow as TData]);
+
+    return function removePendingRow() {
+      pendingRows.current = pendingRows.current.filter((r) => r !== pendingRow);
+      setTableData((prev) => prev.filter((r) => r !== pendingRow));
+    };
+  };
```

Comment:
I cannot, because I make reference equality with `pendingRows` and in `addPendingRow` I create a new object with `fillPendingRow`.

Adding `replacePendingRow` complicate the code for nothing.

>  it can potentially lead to stale state, If the state changes before the operation completes, calling this cleanup function could inadvertently revert some of those changes.

No because a callback in `setTableData`, `prev` is the last value when the function is called.

Reply:
I personally prefer to separate concerns this way you know exactly at first look what each function does, but as you prefer. Yes it's better  like in the example 👍🏼

Reply:
Ok finally I prefer to separate too.